transport.go

```diff
@@ -1,791 +1,148 @@
-
-	shutdown     bool
-	shutdownCh   chan struct{}
-	shutdownLock sync.Mutex
-}
 
-// appendFuture implements raft.AppendFuture.
-// It is used for waiting on a pipelined append
-// entries RPC.
-// Copied from raft future.go because it is private there
-// even though I don't know why
-type appendFuture struct {
-	deferError
-	start time.Time
-	args  *raft.AppendEntriesRequest
-	resp  *raft.AppendEntriesResponse
+	return gostream.Dial(sl.host, pid, RaftProtocol)
 }
 
-func (a *appendFuture) Start() time.Time {
-	return a.start
+func (sl *streamLayer) Accept() (net.Conn, error) {
+	return sl.l.Accept()
 }
 
-func (a *appendFuture) Request() *raft.AppendEntriesRequest {
-	return a.args
+func (sl *streamLayer) Addr() net.Addr {
+	return sl.l.Addr()
 }
 
-func (a *appendFuture) Response() *raft.AppendEntriesResponse {
-	return a.resp
+func (sl *streamLayer) Close() error {
+	return sl.l.Close()
 }
 
-// deferError can be embedded to allow a future
-// to provide an error in the future.
-// Copied from raft future.go
-type deferError struct {
-	err       error
-	errCh     chan error
-	responded bool
+type addrProvider struct {
+	h host.Host
 }
 
-func (d *deferError) init() {
-	d.errCh = make(chan error, 1)
-}
-
-func (d *deferError) Error() error {
-	if d.err != nil {
-		// Note that when we've received a nil error, this
-		// won't trigger, but the channel is closed after
-		// send so we'll still return nil below.
-		return d.err
-	}
-	if d.errCh == nil {
-		panic("waiting for response on nil channel")
-	}
-	d.err = <-d.errCh
-	return d.err
-}
-
-func (d *deferError) respond(err error) {
-	if d.errCh == nil {
-		return
-	}
-	if d.responded {
-		return
-	}
-	d.errCh <- err
-	close(d.errCh)
-	d.responded = true
-}
-
-// newLibp2pPipeline opens a stream to the target, launches a go routine
-// to read responses from that stream continuosly
-func newLibp2pPipeline(tr *Libp2pTransport, target string) (*libp2pPipeline, error) {
-	logger.Debugf("%s: New pipeline to %s", tr.LocalAddr(), target)
-	p, err := peer.IDB58Decode(target)
-	if err != nil {
-		return nil, err
-	}
-
-	stream, err := tr.host.NewStream(tr.ctx, p, RaftPipelineProtocol)
+// ServerAddr takes a raft.ServerID and checks that it is a valid PeerID and
+// that our libp2p host has at least one address to contact it. It then returns
+// the ServerID as ServerAddress. On all other cases it will throw an error.
+func (ap *addrProvider) ServerAddr(id raft.ServerID) (raft.ServerAddress, error) {
+	pid, err := peer.IDB58Decode(string(id))
 	if err != nil {
-		logger.Debugf("%s: %s", tr.LocalAddr(), err)
-		return nil, err
+		return "", fmt.Errorf("bad peer ID: %s", id)
 	}
-	wrap := wrapStream(stream)
-
-	pipeline := &libp2pPipeline{
-		trans: tr,
-		sWrap: wrap,
-
-		doneCh:       make(chan raft.AppendFuture, rpcMaxPipeline),
-		inProgressCh: make(chan *appendFuture, rpcMaxPipeline),
-		shutdownCh:   make(chan struct{}),
+	addrs := ap.h.Peerstore().Addrs(pid)
+	if len(addrs) > 0 {
```

Comment:
Good catch. It did not show up in the tests because Raft falls back on ServerAddress if this fails and it's set to be the same as the ServerID, but it did print some warnings which I was wondering about yesterday...